### ISSUE TITLE
WAL-G xtrabackup-push is slow

### ISSUE THREAD

**Issue by bmday** [wal-g/wal-g#1657]
### Database name
Wal-g mysql version   13b9a8c5        2024.02.20_11:37:02
MySQL 5.7.39-42-log Percona Server
Our database have a 1.1TB of data.
Network сhannel width 10Gbit to backuphost.
SSD U.2 PM9A3

# Issue description

### Describe your problem
We would like to implement wal-g to perform a backup of our database.
For now we are using xtrabackup, which we launch with the following arguments:
`xtrabackup --defaults-extra-file=/root/.my.cnf --no-timestamp --lock-ddl-per-table $DIR/$DT --slave-info --parallel=1 --stream=xbstream | pzstd | ssh ${SFTP} "cat - > ${RPATH}/${FN}"`
This operation performs a backup of a 1.1TB database in 40 minutes.

For testing, we decided to use wal-g, which we compiled ourselves from the master branch dated February 20th.
The application works, but xtrabackup-push takes approximately 2 hours 35 minutes. I am attaching the wal-g configuration in this appeal. We can't understand why xtrabackup-push takes so long.

Our monitoring showed that bandwidth usage during backup via xtrabackup tries to use all the available bandwidth for the network connection.
When we use wal-g with xtrabackup-push, the data transfer speed does not exceed 200 Mbit/s

The backuphost is not busy with other tasks, its network channel is not loaded.

### Please add config and wal-g stdout/stderr logs for debug purpose
```
WARNING: 2024/03/13 04:00:02.091304 WALG_FAILOVER_STORAGES_CACHE_LIFETIME is unknown
WARNING: 2024/03/13 04:00:02.091350 WALG_FAILOVER_STORAGES_CHECK_TIMEOUT is unknown
WARNING: 2024/03/13 04:00:02.091403 We found that some variables in your config file detected as 'Unknown'.
  If this is not right, please create issue https://github.com/wal-g/wal-g/issues/new
DEBUG: 2024/03/13 04:00:02.092966 --- COMPILED ENVIRONMENT VARS ---
        SSH_PORT=22
        SSH_PRIVATE_KEY_PATH=/root/.ssh/id_rsa
        SSH_USERNAME=root
        TOTAL_BG_UPLOADED_LIMIT=32
        WALG_BACKUP_DOWNLOAD_MAX_RETRY=1
        WALG_COMPRESSION_METHOD=zstd
        WALG_DELTA_MAX_STEPS=0
        WALG_DOWNLOAD_CONCURRENCY=10
        WALG_DOWNLOAD_FILE_RETRIES=15
        WALG_ENVELOPE_CACHE_EXPIRATION=0
        WALG_FAILOVER_STORAGES_CACHE_LIFETIME=15m
        WALG_FAILOVER_STORAGES_CHECK_TIMEOUT=30s
        WALG_INTEGRITY_MAX_DELAYED_WALS=0
        WALG_LIBSODIUM_KEY_TRANSFORM=none
        WALG_LOG_LEVEL=DEVEL
        WALG_MYSQL_BACKUP_PREPARE_COMMAND=xtrabackup --prepare --target-dir=/var/lib/mysql/data
        WALG_MYSQL_BINLOG_DST=/var/lib/mysql/data/my-binlogs
        WALG_MYSQL_BINLOG_REPLAY_COMMAND=mysqlbinlog --stop-datetime="$WALG_MYSQL_BINLOG_END_TS" "$WALG_MYSQL_CURRENT_BINLOG" | mysql
        WALG_MYSQL_CHECK_GTIDS=true
        WALG_MYSQL_DATASOURCE_NAME=username:password@tcp(localhost:3306)/db
        WALG_MYSQL_INCREMENTAL_BACKUP_DST=/tmp
        WALG_PREVENT_WAL_OVERWRITE=false
        WALG_SERIALIZER_TYPE=json_default
        WALG_SKIP_REDUNDANT_TARS=false
        WALG_SSH_PREFIX=ssh://backuphost/dbbackup/
        WALG_STORE_ALL_CORRUPT_BLOCKS=false
        WALG_STREAM_CREATE_COMMAND=/usr/bin/xtrabackup --defaults-extra-file=/root/.my.cnf --lock-ddl-per-table --backup --stream=xbstream --datadir=/var/lib/mysql/data --parallel=1 --no-timestamp --slave-info
        WALG_STREAM_RESTORE_COMMAND=xbstream -x -C /var/lib/mysql/data
        WALG_STREAM_SPLITTER_BLOCK_SIZE=1048576
        WALG_TAR_DISABLE_FSYNC=false
        WALG_TAR_SIZE_THRESHOLD=1073741823
        WALG_UPLOAD_CONCURRENCY=24
        WALG_UPLOAD_DISK_CONCURRENCY=2
        WALG_UPLOAD_QUEUE=2
        WALG_UPLOAD_WAL_METADATA=NOMETADATA
        WALG_USE_COPY_COMPOSER=false
        WALG_USE_DATABASE_COMPOSER=false
        WALG_USE_RATING_COMPOSER=false
        WALG_USE_REVERSE_UNPACK=false
        WALG_USE_WAL_DELTA=false
        WALG_VERIFY_PAGE_CHECKSUMS=false
        WALG_WITHOUT_FILES_METADATA=false
DEBUG: 2024/03/13 04:00:02.424169 No crypter has been selected
INFO: 2024/03/13 04:00:03.150209 Full backup requested.
INFO: 2024/03/13 04:00:03.150361 Command to execute: /bin/sh -c /usr/bin/xtrabackup --defaults-extra-file=/root/.my.cnf --lock-ddl-per-table --backup --stream=xbstream --datadir=/var/lib/mysql/data --parallel=1 --no-timestamp --slave-info --extra-lsndir=/tmp/wal-g486079869
INFO: 2024/03/13 06:35:42.109463 FILE PATH: stream_20240313T010003Z/stream.zst
INFO: 2024/03/13 06:35:42.115823 Backup sentinel: {"Tool":"WALG_XTRABACKUP_TOOL","BinLogStart":"mysql-bin.006509","BinLogEnd":"mysql-bin.006509","StartLocalTime":"2024-03-13T04:00:03.150202+03:00","StopLocalTime":"2024-03-13T06:35:42.115543+03:00","UncompressedSize":1093042107405,"CompressedSize":145401587818,"Hostname":"backuphost","ServerUUID":"d4fde908-6f73-11ed-9b46-86a5be8d69f3","ServerVersion":"5.7.39-42-log","IsPermanent":false,"IsIncremental":false,"LSN":39161717493978,"DeltaLSN":0,"DeltaCount":0}
```

![image](https://github.com/wal-g/wal-g/assets/135819656/14075c99-7b42-4eec-9e70-36f5e02b4795)

**Comment by ostinru** [comment]
Hi @bmday!
Basically, with your config wal-g doing following: `xtrabackup | zstd | ssh`. And it works with a pace of the slowest part of pipeline. In my experience, slowest is encryption. In your case, it seems, that it is compression. I suppose, that `pzstd` doing some multi-threading, wal-g don't.

With wal-g we can use SplitStream backup format - it will split single backup stream into N streams, and compress/encrypt them separately. As a result you will have N parts of backup on remote host.
To enable this behaviour you should set:
```
WALG_STREAM_SPLITTER_PARTITIONS = 8
WALG_STREAM_SPLITTER_BLOCK_SIZE = 1048576
```
and probably, run `wal-g backup-push` instead of `xtrabackup-push` (it will do the same things, but won't support diff/incremental backups won't be supported).

NOTE: it will be tricky to restore multi-file backup without wal-g. However we use it in production for long time with no issues.

---

In `xtrabackup-push` I would like to support `--fifo-streams`[1][2], that will make SplitStream obsolete, and I don't think that we should support SplitStream with `xtrabackup-push`.

[1] https://docs.percona.com/percona-xtrabackup/8.0/xbcloud-binary-fifo-datasink.html
[2] https://github.com/wal-g/wal-g/pull/1650

**Comment by bmday** [comment]
@ostinru thank you for your reply.
We also tried backup-push instead of xtrabackup-push, but it did not affect the execution time.

It seems to me that you are right when you say that the reason may be zstd. We use pzstd, which by default compresses into 4 threads, which correlates with execution time, it is 4 times less.

How difficult do you think it is to implement multi-threaded compression in wal-g? For example, there would be a variable WALG_<compression_method>_THREADS=4, which would control the number of compression threads for <compression_method>. At least for zstd.

**Comment by ostinru** [comment]
> but it did not affect the execution time.

You can use multiple threads with `WALG_STREAM_SPLITTER_PARTITIONS` - it will split single backup stream and compress each part in separate thread.

> How difficult do you think it is to implement multi-threaded compression in wal-g?

In the future `xtrabackup --fifo-streams` should produce N separate streams that we will be possible to compress/encode separately.

About general multi-threaded compression/encryption:
* If we trust wal-g in restore process - then it is already done with `WALG_STREAM_SPLITTER_PARTITIONS` and well tested.
* If we want to be able to restore backup with simple bash pipeline - then I am pretty sure that it is not solvable in _general_ case: not all algorithms we support can run in parallel on single backup stream.

**Comment by bmday** [comment]
Thank you. We have received the answer to our question.
